src/nvpair.c

```diff
@@ -205,8 +226,10 @@ nvpair_remove_nvlist_array(nvpair_t *nvp)
 	/* XXX: DECONST is bad, mkay? */
 	nvlarray = __DECONST(nvlist_t **,
 	    nvpair_get_nvlist_array(nvp, &count));
-	for (i = 0; i < count; i++)
+	for (i = 0; i < count; i++) {
 		nvlist_set_array_next(nvlarray[i], NULL);
+		nvlist_set_parent(nvlarray[i], NULL);
```

Comment:
Very very nice catch!
Could you also add a test case for this?

Reply:
I could, but unfortunately I do not have a spare FreeBSD system with the C++ ATF stuff (and, interestingly, ATF in NetBSD does not seem to be compatible).  It is indirectly covered by the last case in `nvlist_append_test.c` -- but yes, we should have a proper test..